// libp2p-topology/dist/src/index.js
const noop = () => { };
const topologySymbol = Symbol.for('@libp2p/js-interfaces/topology');
export class Topology {
    constructor(options) {
        this.min = options.min ?? 0;
        this.max = options.max ?? Infinity;
        this.peers = new Set();
        this._onConnect = options.handlers?.onConnect == null ? noop : options.handlers?.onConnect;
        this._onDisconnect = options.handlers?.onDisconnect == null ? noop : options.handlers?.onDisconnect;
    }
    get [Symbol.toStringTag]() {
        return 'Topology';
    }
    get [topologySymbol]() {
        return true;
    }
    /**
     * Checks if the given value is a Topology instance
     */
    static isTopology(other) {
        return topologySymbol in other;
    }
    set registrar(registrar) {
        this._registrar = registrar;
    }
    get registrar() {
        return this._registrar;
    }
    /**
     * Notify about peer disconnected event
     */
    disconnect(peerId) {
        this._onDisconnect(peerId);
    }
}
