// libp2p-peer-id-factory/dist/src/index.js
import { keys } from 'libp2p-crypto';
import { fromString as uint8ArrayFromString } from 'uint8arrays/from-string';
import { PeerId } from 'libp2p-peer-id';
import { PeerIdProto } from './proto.js';
export const createEd25519PeerId = async () => {
    const key = await keys.generateKeyPair('Ed25519');
    const id = await createFromPrivKey(key);
    if (id.type === 'Ed25519') {
        return id;
    }
    throw new Error(`Generated unexpected PeerId type "${id.type}"`);
};
export const createSecp256k1PeerId = async () => {
    const key = await keys.generateKeyPair('secp256k1');
    const id = await createFromPrivKey(key);
    if (id.type === 'secp256k1') {
        return id;
    }
    throw new Error(`Generated unexpected PeerId type "${id.type}"`);
};
export const createRSAPeerId = async (opts) => {
    const key = await keys.generateKeyPair('RSA', opts?.bits ?? 2048);
    const id = await createFromPrivKey(key);
    if (id.type === 'RSA') {
        return id;
    }
    throw new Error(`Generated unexpected PeerId type "${id.type}"`);
};
export async function createFromPubKey(publicKey) {
    return await PeerId.fromKeys(keys.marshalPublicKey(publicKey));
}
export async function createFromPrivKey(privateKey) {
    return await PeerId.fromKeys(keys.marshalPublicKey(privateKey.public), keys.marshalPrivateKey(privateKey));
}
export function exportToProtobuf(peerId, excludePrivateKey) {
    return PeerIdProto.encode({
        id: peerId.multihash.bytes,
        pubKey: peerId.publicKey,
        privKey: excludePrivateKey === true || peerId.privateKey == null ? undefined : peerId.privateKey
    }).finish();
}
export async function createFromProtobuf(buf) {
    const { id, privKey, pubKey } = PeerIdProto.decode(buf);
    return await createFromParts(id, privKey, pubKey);
}
export async function createFromJSON(obj) {
    return await createFromParts(uint8ArrayFromString(obj.id, 'base58btc'), obj.privKey != null ? uint8ArrayFromString(obj.privKey, 'base64pad') : undefined, obj.pubKey != null ? uint8ArrayFromString(obj.pubKey, 'base64pad') : undefined);
}
async function createFromParts(multihash, privKey, pubKey) {
    if (privKey != null) {
        const key = await keys.unmarshalPrivateKey(privKey);
        return await createFromPrivKey(key);
    }
    else if (pubKey != null) {
        const key = await keys.unmarshalPublicKey(pubKey);
        return await createFromPubKey(key);
    }
    return PeerId.fromBytes(multihash);
}
